evaluator/src/main/resources/rules/osadl.rules.kts

```diff
@@ -0,0 +1,114 @@
+    UNKNOWN;
+
+    companion object {
+        fun fromString(value: String): Compatibility =
+            when (value) {
+                "", "Same" -> INHERENT
+                "Yes" -> YES
+                "No" -> NO
+                "?", "Unknown", "Check dependency", "Dep." -> UNKNOWN
+                else -> throw IllegalArgumentException("Unhandled compatibility '$value' found.")
+            }
+    }
+}
+
+val matrix = mutableMapOf<Pair<String, String>, Compatibility>()
+
+javaClass.getResourceAsStream("/rules/osadl-matrix.csv").bufferedReader().useLines { lines ->
+    val rowIterator = lines.iterator()
+
+    // The first row contains the header with names of the inbound licenses.
+    val columnHeaders = rowIterator.next().split(',')
+        // Drop the surrounding "Compatibility*" cells from the header row.
+        .drop(1).dropLast(1)
+
+    while (rowIterator.hasNext()) {
+        val cellIterator = rowIterator.next().split(',').iterator()
+
+        // The first column contains the header with names of the outbound licenses.
+        val inbound = cellIterator.next()
+
+        columnHeaders.forEach { outbound ->
+            val value = cellIterator.next()
+            matrix[inbound to outbound] = Compatibility.fromString(value)
+        }
+    }
+}
+
+fun PackageRule.LicenseRule.isOsadlMatrixNonCompliant(inbound: SpdxSingleLicenseExpression) =
+    object : RuleMatcher {
+        override val description = "isOsadlMatrixCompliant($license)"
+
+        override fun matches(): Boolean {
+            // Be conservative and use the simple license string without the exception string for lookup. Eventually
+            // resulting false-positives are better than an incompatibility to be missed.
+            val outbound = license.simpleLicense()
+            return matrix[inbound.simpleLicense() to outbound] == Compatibility.NO
+        }
+    }
+
+val ruleSet = ruleSet(ortResult, licenseInfoResolver) {
+    // Define a rule that is executed for each dependency.
+    dependencyRule("OSADL compliance matrix") {
+        // Requirements for the rule to trigger a violation.
+        require {
+            -isExcluded()
+        }
+
+        // Define a rule that is executed for each license of the dependency.
+        val licenseView = LicenseView.CONCLUDED_OR_DECLARED_AND_DETECTED
+        licenseRule("OSADL compliance matrix", licenseView) {
+            // Use all licenses of all direct dependencies as inbound licenses.
+            directDependencies().forEach { dependency ->
+                val resolveLicenseInfo = licenseInfoResolver.resolveLicenseInfo(dependency.id).filter(licenseView)
+                val inboundLicenses = resolveLicenseInfo.licenses.map { it.license }
+
+                inboundLicenses.forEach { inboundLicense ->
+                    // Requirements for the rule to trigger a violation.
+                    require {
+                        +isOsadlMatrixNonCompliant(inboundLicense)
+                    }
```

Comment:
I would be surprised if this works as intended. The requirements are not magically only applied to the code block below, instead you add a new matcher to the dependency rule with each iteration of the loop, and they are all concatenated with `and`, and the code below also adds a new violation for each iteration. So if all matchers match you will get all violations, but if a single matcher does not match you will get no violations. To see this in the tests you would have to use test data with multiple dependencies on the same level.

If you want to use loops inside the rule simply use `if` instead of `require`.